python/rascaline/rascaline/utils/splines.py

```diff
@@ -0,0 +1,371 @@
+"""
+Splined radial integrals
+========================
+
+Classes for generating spline points which can be used as input for the for the radial
+integral hyper paramater option. For an complete example for the LE how to use the
+classes see :ref:`example-splines`.
+
+.. autoclass:: rascaline.utils.SplinerBase
+    :members:
+    :show-inheritance:
+
+.. autoclass:: rascaline.utils.SplinerFromFunction
+    :members:
+    :show-inheritance:
+
+
+"""
+
+from abc import ABC, abstractmethod
+from typing import Callable, Dict, Optional
+
+import numpy as np
+
+
+class SplinerBase(ABC):
+    """Base class for splining arbitrary radial integrals.
+
+    If ``_radial_integral_derivative`` is not implemented in a child class it will
+    computed based on finite differences.
+
+    :parameter cutoff: Spherical cutoff for the radial basis
+    :parameter max_angular: number of radial components
+    :parameter max_radial: number of angular components
+    :parameter spline_cutoff: cutoff radius for the spline interpolation. This is also
+        the maximal value that can be interpolated.
+    :parameter accuracy: accuracy of the numerical integration and the splining.
+        Accuracy is reached when either the mean absolute error or the mean relative
+        error gets below the ``accuracy`` threshold.
+    """
+
+    def __init__(
+        self,
+        cutoff: float,
+        max_radial: int,
+        max_angular: int,
+        spline_cutoff: float,
+        accuracy: float,
+    ):
+        self.cutoff = cutoff
+        self.max_radial = max_radial
+        self.max_angular = max_angular
+        self.spline_cutoff = spline_cutoff
+        self.accuracy = accuracy
+
+    @abstractmethod
+    def _radial_integral(self, n: int, ell: int, positions: np.ndarray) -> np.ndarray:
+        """Method calculating the radial integral."""
+        ...
+
+    def _radial_integral_derivative(
+        self, n: int, ell: int, positions: np.ndarray
+    ) -> np.ndarray:
+        """Method calculating the derivatice of the radial integral."""
+
+        displacement = 1e-6
+
+        radial_integral_pos = self._radial_integral(
+            n, ell, positions + displacement / 2
+        )
+        radial_integral_neg = self._radial_integral(
+            n, ell, positions - displacement / 2
+        )
```

Comment:
In my experience, this can cause problems when evaluating derivatives at the edges of the domain (i.e., at 0 and r_cut) because the function might not be defined outside of them. However, you could also make the case that it's up to the user to provide a custom derivative function in such a situation.

Reply:
Yes, I agree. I think the numerical derivative is really the fallback solution if you can not provide a analytical derivative. We can clarify this a bit more in the doc string.

We can also improve the code here.